shinysdr/plugins/limesdr.py

```diff
@@ -0,0 +1,306 @@
+# Copyright 2013, 2014, 2015, 2016, 2017 Kevin Reid <kpreid@switchb.org>
+#
+# This file is part of ShinySDR.
+# 
+# ShinySDR is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+# 
+# ShinySDR is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+# 
+# You should have received a copy of the GNU General Public License
+# along with ShinySDR.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import absolute_import, division, unicode_literals
+
+from zope.interface import implementer  # available via Twisted
+
+from gnuradio import gr
+from gnuradio import blocks
+
+import limesdr
+
+from shinysdr.devices import Device, IRXDriver
+from shinysdr.signals import SignalType
+from shinysdr.types import EnumT, RangeT
+from shinysdr import units
+from shinysdr.values import ExportedState, LooseCell, exported_value, nullExportedState, setter
+
+
+__all__ = []
+
+
+# TODO: Support both channels on LimeSDRUSB
+ch = 0  # single channel number used
+
+
+# Constants, not exported by gr-limesdr
+LimeSDRMini = 1
+LimeSDRUSB = 2
+SISO = 1
+MIMO = 2
+A = 0
+B = 1
+LNANONE = 0
+LNAH = 1
+LNAL = 2
+LNAW = 3
+UPCONVERT = 0
+DOWNCONVERT = 1
+
+
+# default tune_delay value
+DEFAULT_DELAY = 0.07
+
+
+class _LimeSDRTuning(object):
```

Comment:
They're full duplex but there are still separate source and sink blocks. I think I still need to have this object structure because the RX and TX drivers are separate hier blocks, no?

Reply:
The reason that the OsmoSDR code is laid out as it is is because it is critical to have prompt garbage collection of the blocks when TX/RX switching, so circular references need to be avoided, and because the blocks need to be recreated so the state needs to be saved. (I think the current code fails to do all that well enough, but that's the motivation.)

You're probably right that this is a reasonable structure just due to there being two objects.